### DISCUSSION TITLE
Extra card icons and HTML elements in description text for payment gateways.

### DISCUSSION THREAD

**Discussion by AaronBowie** [woocommerce/woocommerce#45016] [WooCommerce, Blocks, and Site Editing]
Hi all.

We have been adding “basic” support for the new checkout block for our gateway plugins, but there are a few things I can't work out, and I have spent a few days going through the GitHub repo and issues. Sadly, I have unable to find some solutions for these issues we are having.

I've attached a screenshot of how our plugins work in the normal checkout option.

First issue, showing multiple card icons, the merchant is able to select which card icons are shown in the settings page. Right now, we only seem to be able to add one image next to the gateway name.

Second issue, we are unable to enter in HTML in the gateway description. For us, we use this to display some information while in test mode, so the merchants can find a test card to use and also a quick link to the docs. We have seen, in the past, merchants also editing the description text using HTML tags.

For now, we have stripped HTML from the description, and have an icon with multiple cards, but I know that our customers will be asking for the above once we get more using the new checkout blocks.

![image (77)](https://user-images.githubusercontent.com/6169614/197756777-2bcd304d-1ff8-43c8-baab-390f0f100f9f.png)
![screenshot-oppcopyandpay-local-checkout-blocks-1666655033375 (1)](https://user-images.githubusercontent.com/6169614/197756804-b7f2ba2a-1699-4782-b5ee-aa070dfb9b23.png)

**Comment by senadir** [comment]
Both the payment method title and content area react nodes, have you attempted to add your render logic there? when you call `registerPaymentMethod`?

**Comment by AaronBowie** [comment]
Hi @senadir,

Many thanks for the quick reply. 

React is not something we are strong with, so we are trying to pick this up quickly.

**Comment by senadir** [comment]
Unfortunately, your use case is something to be served with React, this is how your code will probably look like:

```jsx
const settings = getSetting( 'yourpaymentname_data', {} );

/**
 * Content component
 */
const Content = () => {
	const isTestMode = settings.testMode;
	return (
		<p>
			{ decodeEntities( settings.description ) }
			{ isTestMode &&
				"You're in test mode, you can add extra informations here" }
		</p>
	);
};

const Label = () => {
	const cardIcons = settings.cardIcons;
	return (
		<div className="my-payment-method-title-wrapper">
			<span className="my-payment-method-title">Plugin Name</span>
			{ cardIcons.map( ( icon ) => (
				<img src={ icon.src } alt={ decodeEntities( icon.title ) } />
			) ) }
		</div>
	);
};
const paypalPaymentMethod = {
	name: PAYMENT_METHOD_NAME,
	label: <Label />,
	content: <Content />,
	edit: <Content />,
	canMakePayment: () => true,
	ariaLabel: decodeEntities( settings.title ),
	supports: {
		features: settings.supports ?? [],
	},
};

registerPaymentMethod( paypalPaymentMethod );
```


In your `get_payment_method_data` function, you can register an array of icons and their titles, and a boolean for test mode.

If you don't have a JSX/React build process, you can use this directly (it's already transformed):

```js
const settings = getSetting('yourpaymentname_data', {});

/**
 * Content component
 */
const Content = () => {
  const isTestMode = settings.testMode;
  return createElement("p", null, decodeEntities(settings.description), isTestMode && "You're in test mode, you can add extra informations here");
};
const Label = () => {
  const cardIcons = settings.cardIcons;
  return createElement("div", {
    className: "my-payment-method-title-wrapper"
  }, createElement("span", {
    className: "my-payment-method-title"
  }, "Plugin Name"), cardIcons.map(icon => createElement("img", {
    src: icon.src,
    alt: decodeEntities(icon.title)
  })));
};
const paypalPaymentMethod = {
  name: PAYMENT_METHOD_NAME,
  label: createElement(Label, null),
  content: createElement(Content, null),
  edit: createElement(Content, null),
  canMakePayment: () => true,
  ariaLabel: decodeEntities(settings.title),
};
registerPaymentMethod(paypalPaymentMethod);
```

You can [transform code here](https://babeljs.io/repl/#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.6&spec=false&loose=false&code_lz=PQKhAIAECsGcA9wGMBOBTAhgFzQUQDZoC2aAdluCMAFBID2psFA3uOgOYCWTaKAChgCeJcgFk0WABZ0AJuAC-4ALzgA7kgB06gEL46SANawASmi5MUggNy0GTcK1SYcBYmQqKVqgA4a0hESwbW0YKWAksTlJ2WGVwdgkAZQio9gAKcAByQToAVxRvIUDSDBIAfRlsDEyAGgdFAEpg0BBqSnAAYQYccmQ6Im8GdzaqEPsu8nc4tIblAD4HagBIelDwbgAVNCZRWTQ48KxI6NgNHB29myX0LHzScDTlpYAebzmnpdYZNHpv3HJOJFthlDscYhpvrBUJxvJEGOBZvIPqxNtssLtvuAAGRYj5LABEAE08pl0Ot7ucKEQ9nUcrlkBh7hgZHI0PAsCgMOSAGZ0FBEbCcOzgSS8ND4hRPZ7AN7LJrUeTBVb2AAyGAARv5prMlAtmMtlRQkBgUDIAJLKg4pE4aY2mi12K43O4PKUyTgAN2Q-AwsFgADlSmglPiiIIALSFYTucMkKSycORLCEcOqTneby8fHvJa556wQr3JA-v2BkghsORooxuPSGSJwGEbN8fC5Lj3MtoaUFxk53OOE3m5UaAXeDIZTirBHzV25vOcIjscCwFBIJQo1YaFdIBTgDD4LDr8DfX54AFA2ATzdJwjTxTAPtLWaIqXAd0enPyxVjChRwr4ARozECQ6zifUlhKEgAC5wD4ABBQlRFwf0NjKJCNgACQAeQAETKf04KQmplh9TV8Bg541TI8AH2IlZuncCiJh6ChaOWNB3SwJiGN6NiVkZUQMAMNBAMCGCZhnDlcjQOiTU4DAqP8GCTz2f4k04YFl2tcEb32Bo6NgXIMz5LBYBg8Clm5Zx8m2GDQVSU5DOMlBTPAAB-NzwAAbQAXTo-RiO_agOG4HB-GrYD4xkDI_33UT3HEKKESsIA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-2&prettier=false&targets=&version=7.19.6&externalPlugins=&assumptions=%7B%7D) directly if you want.

**Comment by AaronBowie** [comment]
Many thanks @senadir, that was super helpful! 

![Screenshot 2022-10-25 154129](https://user-images.githubusercontent.com/6169614/197804701-b5aeed22-cdda-44b3-9958-a5ae9b9fae68.png)
